scanner/src/main/kotlin/experimental/WorkingTreeCache.kt

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2021 Bosch.IO GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.experimental
+
+import java.io.IOException
+
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+import org.ossreviewtoolkit.downloader.VersionControlSystem
+import org.ossreviewtoolkit.downloader.WorkingTree
+import org.ossreviewtoolkit.model.VcsInfo
+import org.ossreviewtoolkit.utils.common.safeDeleteRecursively
+import org.ossreviewtoolkit.utils.core.createOrtTempDir
+
+/**
+ * A cache for [VCS][VersionControlSystem] [WorkingTree]s that manages one working tree for each triple of
+ * [VcsInfo.type], [VcsInfo.url], and [VcsInfo.path].
+ */
+interface WorkingTreeCache {
+    /**
+     * Apply the [block] to the working tree defined by [vcsInfo]. The [VcsInfo.revision] is ignored during
+     * initialization of the working tree, it is the caller's responsibility to update the working tree to the correct
+     * revision. It is ensured that only one [block] at a time can access a single working tree, separate working trees
+     * can be accessed in parallel.
+     */
+    suspend fun <T> use(vcsInfo: VcsInfo, block: (VersionControlSystem, WorkingTree) -> T): T
+
+    /**
+     * Clear all cached working trees from the file system. The function waits for all currently running operations on
+     * the working trees to finish before deleting them.
+     */
+    suspend fun clear()
+}
+
+class DefaultWorkingTreeCache : WorkingTreeCache {
+    private val mutex = Mutex()
+    private val workingTreeMutexes = mutableMapOf<String, Mutex>()
+    private val workingTrees = mutableMapOf<String, WorkingTree>()
+
+    override suspend fun <T> use(vcsInfo: VcsInfo, block: (VersionControlSystem, WorkingTree) -> T): T =
+        getWorkingTreeMutex(vcsInfo).withLock {
+            val vcs = getVcs(vcsInfo)
+            val workingTree = getWorkingTree(vcsInfo)
+
+            block(vcs, workingTree)
```

Comment:
1. `use()` first aquires the `workingTreeMutex` and then the `mutex`
2. clear() first aquires the `mutex` and then each `workingTreeMutex`

...so this fullfills the necessary precondition for a deadlock. Any way to avoid that?

Reply:
No, `use` only locks with `mutex` while acquiring the `workingTreeMutex` in `getWorkingTreeMutex()`. When using the `workingTreeMutex` the lock on `mutex` is already released.

Reply:
Oh, your're right. I missed the release of the lock.